src/ber/util.rs

```diff
@@ -0,0 +1,140 @@
+use super::{DeserializeError, Length, Serialize, Tag};
+use tinyvec::TinyVec;
+
+pub trait VecExt {
+    fn write(&mut self, slice: &[u8]);
+    fn write_byte(&mut self, byte: u8);
+}
+
+impl VecExt for Vec<u8> {
+    fn write(&mut self, slice: &[u8]) {
+        self.extend_from_slice(slice);
+    }
+    fn write_byte(&mut self, byte: u8) {
+        self.push(byte);
+    }
+}
+
+impl VecExt for TinyVec<[u8; 32]> {
+    fn write(&mut self, slice: &[u8]) {
+        self.extend_from_slice(slice);
+    }
+    fn write_byte(&mut self, byte: u8) {
+        self.push(byte);
+    }
+}
+
+pub trait ReadExt {
+    fn peek(&mut self) -> Result<u8, DeserializeError>;
+    fn byte(&mut self) -> Result<u8, DeserializeError>;
+    fn uint(&mut self, size: Length) -> Result<u64, DeserializeError>;
+    fn int(&mut self, size: Length) -> Result<i64, DeserializeError>;
+    fn slice(&mut self, size: Length) -> Result<&[u8], DeserializeError>;
+    fn tag(&mut self, expected: Tag) -> Result<Tag, DeserializeError>;
+    fn peek_tag(&mut self, expected: Tag) -> Result<Tag, DeserializeError>;
+}
+
+impl ReadExt for &'_ [u8] {
+    fn peek(&mut self) -> Result<u8, DeserializeError> {
+        Ok(*self.get(0).ok_or(DeserializeError::BufferTooShort)?)
+    }
+
+    fn byte(&mut self) -> Result<u8, DeserializeError> {
+        let byte = *self.get(0).ok_or(DeserializeError::BufferTooShort)?;
+        *self = &self[1..];
+        Ok(byte)
+    }
+
+    fn uint(&mut self, size: Length) -> Result<u64, DeserializeError> {
```

Comment:
if we are taking in Length and converting to `usize`, why not accept a `usize` directly?

Reply:
Because `usize`'s size is platform dependent

Reply:
Realistically though anything greater than u16 probably will never happen, I'll give that some thought and come back to it sometime to see if the change is worth it